lnrpc/routerrpc/forward_interceptor.go

```diff
@@ -0,0 +1,224 @@
+			}
+		case resolution := <-resolutionRequests:
+			log.Tracef("resolving intercepted packet %v", resolution)
+			// in case we couldn't resolve we just add a log line since this
+			// does not indicate on any connection problem.
+			if err := r.resolveFromClient(resolution); err != nil {
+				log.Warnf("client resolution of intercepted "+
+					"packet failed %v", err)
+			}
+		case err := <-errChan:
+			log.Errorf("Received an error: %v, shutting down", err)
+			return err
+		case <-r.server.quit:
+			return nil
+		}
+	}
+}
+
+// onIntercept is the function that is called by the switch for every forwarded
+// packet. Our interceptor makes sure we hold the packet and then signal to the
+// main loop to handle the packet. We only return true if we were able
+// to deliver the packet to the main loop.
+func (r *forwardInterceptor) onIntercept(p htlcswitch.InterceptedForward) bool {
+	select {
+	case r.intercepted <- p:
+		return true
+	case <-r.quit:
+		return false
+	case <-r.server.quit:
+		return false
+	}
+}
+
+func (r *forwardInterceptor) readClientResponses(
+	resolutionChan chan *ForwardHtlcInterceptResponse, errChan chan error) {
+
+	defer r.wg.Done()
+	for {
+		resp, err := r.stream.Recv()
+		if err != nil {
+			errChan <- err
+			return
+		}
+
+		// Now that we have the response from the RPC client, send it to
+		// the responses chan.
+		select {
+		case resolutionChan <- resp:
+		case <-r.quit:
+			return
+		case <-r.server.quit:
+			return
+		}
+	}
+}
+
+// holdAndforwardToClient forwards the intercepted htlc to the client.
+func (r *forwardInterceptor) holdAndforwardToClient(
+	forward htlcswitch.InterceptedForward) error {
+
+	htlc := forward.Packet()
+	inKey := forward.CircuitKey()
+
+	// First hold the forward, then send to client.
+	r.holdForwards[forward.CircuitKey()] = forward
+	interceptionRequest := &ForwardHtlcInterceptRequest{
+		IncomingCircuitKey: &CircuitKey{
+			ChanId: inKey.ChanID.ToUint64(),
+			HtlcId: inKey.HtlcID,
+		},
+		HtlcPaymentHash: htlc.PaymentHash[:],
+		AmountMsat:      uint64(htlc.Amount),
+		Expiry:          htlc.Expiry,
+	}
+
+	return r.stream.Send(interceptionRequest)
+}
+
+// resolveFromClient handles a resolution arrived from the client.
+func (r *forwardInterceptor) resolveFromClient(
+	in *ForwardHtlcInterceptResponse) error {
+
+	circuitKey := channeldb.CircuitKey{
+		ChanID: lnwire.NewShortChanIDFromInt(in.IncomingCircuitKey.ChanId),
+		HtlcID: in.IncomingCircuitKey.HtlcId,
+	}
+	var interceptedForward htlcswitch.InterceptedForward
+	interceptedForward, ok := r.holdForwards[circuitKey]
+	if !ok {
+		return ErrFwdNotExists
+	}
+	delete(r.holdForwards, circuitKey)
+
+	switch in.Action {
+	case ResolveHoldForwardAction_RESUME:
+		return interceptedForward.Resume()
+	case ResolveHoldForwardAction_FAIL:
+		return interceptedForward.Fail()
+	case ResolveHoldForwardAction_SETTLE:
+		if in.Preimage == nil {
+			return ErrMissingPreimage
+		}
+		preimage, err := lntypes.MakePreimage(in.Preimage)
```

Comment:
We should catch an _invalid_ pre-image for the payment hash here early (so verify the pay hash and preimage match up). Otherwise, a user could mess up and cause a channel force close since that's considered a dire offense atm.

Reply:
Done.